src/foam/nanos/crunch/connection/CapabilityPayloadDAO.js

```diff
@@ -0,0 +1,147 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.crunch.connection',
+  name: 'CapabilityPayloadDAO',
+  extends: 'foam.dao.ProxyDAO',
+  flags: ['java'],
+
+  javaImports: [
+    'foam.dao.*',
+    'foam.core.FObject',
+    'foam.core.X',
+    'foam.dao.DAO',
+    'foam.dao.Sink',
+    'foam.dao.ArraySink',
+    'foam.mlang.predicate.Predicate',
+    'foam.mlang.order.Comparator',
+    'foam.nanos.auth.User',
+    'foam.util.SafetyUtil',
+    'foam.nanos.crunch.Capability',
+    'foam.nanos.crunch.CrunchService',
+    'foam.nanos.crunch.connection.CapabilityPayload',
+    'java.util.ArrayList',
+    'java.util.List'
+  ],
+
+  documentation: `
+    CapabilityPayload details the data objects needed and grant path for a specific capability
+  `,
+
+  axioms: [
+    {
+      name: 'javaExtras',
+      buildJavaClass: function(cls) {
+        cls.extras.push(`
+          public CapabilityPayloadDAO(X x, DAO delegate) {
+            setX(x);
+            setDelegate(delegate);
+          } 
+        `);
+      }
+    }
+  ],
+
+  methods: [
+    {
+      name: 'find_',
+      javaCode: `
+        if ( id instanceof CapabilityPayload ) return super.find_(x, ((CapabilityPayload) id).getId());
+
+        DAO localCapabilityDAO = (DAO) x.get("localCapabilityDAO");
+
+        String idToString = (String) id;
+
+        // TODO: IN deployment we won't even be storing CapabilityPayload
+        if ( localCapabilityDAO.find(idToString) != null ){
+          CrunchService crunchService = (CrunchService) x.get("crunchService");
+
+          List grantPath = crunchService.getGrantPath(x, idToString);
+          List<Object> capabilityDataObjects = new ArrayList<>();
+          List<String> capabilities = new ArrayList<>();
+          for ( Object item : grantPath ) {
+            // TODO: Verify that MinMaxCapability lists can be ignored here
+            if ( item instanceof Capability ) {
+              Capability cap = (Capability) item;
+              if ( cap.getOf() != null ){
+                FObject newOfInstance;
+                try {
+                  newOfInstance = (FObject) cap.getOf().newInstance();
+                } catch (Exception e){
+                  throw new RuntimeException(e);
+                }
+                capabilityDataObjects.add(newOfInstance);
```

Comment:
the indexes of this array have to correspond exactly to the indexes of the capabilities array in order for the put method to work properly, or else you'll be putting the wrong data object into the wrong UCJ

Reply:
Re new fix